design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+[`range`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#range-trait
+[`required`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#required-trait
+[`uniqueItems`]: https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html#uniqueItems-trait
+[`awslabs/smithy`]: https://github.com/awslabs/smithy
+
+Implementation
+--------------
+
+This section addresses how to implement and enforce the `length`, `pattern`,
+`range`, and `uniqueItems` traits. We will use the `length` trait applied to a
+`string` shape as a running example. The implementation of this trait mostly
+carries over to the other three.
+
+### Example implementation for the `length` trait
+
+Consider the following Smithy model:
+
+```smithy
+@length(min: 1, max: 69)
+string NiceString
+```
+
+The central idea to the implementation of constraint traits is: [parse, don't
+validate]. Instead of code-generating a Rust `String` to represent `NiceString`
+values and perform the _validation_ at request deserialization, we can
+[leverage Rust's type system to guarantee domain invariants]. We can generate a
+wrapper [tuple struct] that _parses_ the string's value and is "tight" in the
+set of values it can accept:
+
+```rust
+pub struct NiceString(String);
+
+impl NiceString {
+    pub fn parse(inner: String) -> Result<Self, NiceStringValidationError> {
+        Self::try_from(inner)
+    }
+}
+
+impl TryFrom<String> for NiceString {
+    type Error = NiceStringValidationError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        let num_code_points = value.chars().count();
+        if 1 <= num_code_points && num_code_points <= 69 {
+            Ok(Self(value))
+        } else {
+            Err(NiceStringValidationError::LengthViolation(num_code_points))
+        }
+    }
+}
+```
+
+(Note that we're using the _linear_ time check `chars().count()` instead of
+`len()` on the input value, since the Smithy specification says the `length`
+trait counts [the number of _Unicode code points_ when applied to string
+shapes].)
+
+The goal is to enforce, at the type-system level, that these constrained
+structs always hold valid data. It should be impossible for the service
+implementer, without resorting to `unsafe` Rust, to construct a `NiceString`
+that violates the model. The actual check is performed in the implementation of
+[`TryFrom`]`<InnerType>` for the generated struct, which makes it convenient to use
+the [`?` operator for error propagation]. Each constrained struct will have a
+related `std::error::Error` enum type to signal the _first_ parsing failure,
+with one enum variant per applied constraint trait:
+
+```rust
+pub enum NiceStringValidationError {
+    /// Validation error holding the number of Unicode code points found, when a value between `1` and
+    /// `69` (inclusive) was expected.
+    LengthViolation(usize),
```

Comment:
It might be too much information, but I always prefer errors that explain not only what the invalid value was, but what counts as a valid value.

Reply:
Nevermind, I see in the next paragraph that you say you'll do this.